reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,84 +83,135 @@ object StreamSubscriber {
         }
         case OnFinalize => {
-          case WaitingOnUpstream(sub, r) =>
-            DownstreamCancellation -> (F.delay(sub.cancel) >> r.complete(None.asRight))
-          case Idle(sub) => DownstreamCancellation -> F.delay(sub.cancel)
-          case o         => o -> F.unit
+          case WaitingOnUpstream(sub) =>
+            (F.delay(sub.cancel) >> q.enqueue1(None.asRight)).as(DownstreamCancellation)
+          case Idle(sub) =>
+            F.delay(sub.cancel).as(DownstreamCancellation)
+          case Receiving(sub) =>
+            F.delay(sub.cancel).as(DownstreamCancellation)
+          case o =>
+            o.pure[F]
         }
-        case OnDequeue(r) => {
-          case Uninitialized          => RequestBeforeSubscription(r) -> F.unit
-          case Idle(sub)              => WaitingOnUpstream(sub, r) -> F.delay(sub.request(1))
-          case err @ UpstreamError(e) => err -> r.complete(e.asLeft)
-          case UpstreamCompletion     => UpstreamCompletion -> r.complete(None.asRight)
-          case o                      => o -> r.complete((new Error(s"received request in invalid state [$o]")).asLeft)
+        case OnDequeue => {
+          case Uninitialized =>
+            (RequestBeforeSubscription: State).pure[F]
+          case Receiving(sub) =>
+            (Receiving(sub): State).pure[F]
+          case err @ UpstreamError(e) =>
+            (q.enqueue1(e.asLeft) >> q.enqueue1(None.asRight)).as(err)
+          case Idle(sub) =>
+            F.delay(sub.request(maxDemand)).as(WaitingOnUpstream(sub)) // request on first dequeue
+          case UpstreamCompletion =>
+            q.enqueue1(None.asRight).as(UpstreamCompletion)
+          case st => st.pure[F]
         }
       }
 
-    Ref.of[F, State](Uninitialized).map { ref =>
+    Ref.of[F, State](Uninitialized) map { ref =>
       new FSM[F, A] {
-        def nextState(in: Input): F[Unit] = ref.modify(step(in)).flatten
-        def onSubscribe(s: Subscription): F[Unit] = nextState(OnSubscribe(s))
-        def onNext(a: A): F[Unit] = nextState(OnNext(a))
-        def onError(t: Throwable): F[Unit] = nextState(OnError(t))
-        def onComplete: F[Unit] = nextState(OnComplete)
-        def onFinalize: F[Unit] = nextState(OnFinalize)
-        def dequeue1: F[Either[Throwable, Option[A]]] =
-          Deferred[F, Out].flatMap(p => ref.modify(step(OnDequeue(p))).flatten >> p.get)
+        def nextState(in: Input): F[Unit] =
+          ref.get
+            .flatMap(step(in))
+            .flatMap(ref.set)
+        def onSubscribe(s: Subscription): F[Unit] =
+          nextState(OnSubscribe(s))
+        def onNext(a: A): F[Unit] =
+          nextState(OnNext(a))
+        def onError(t: Throwable): F[Unit] =
+          nextState(OnError(t))
+        def onComplete: F[Unit] =
+          nextState(OnComplete)
+        def onFinalize: F[Unit] =
+          nextState(OnFinalize)
+        def dequeueChunk1: F[Chunk[Either[Throwable, Option[A]]]] =
```

Comment:
If this is called prior to `onSubscribe`, we'll deadlock, which could happen if the producer asynchronously processes a `Producer#subscribe`.

1. `dequeueChunk1` reads the `Ref`, sees the state is `Uninitialized`
2. Computes the next state which is `RequestBeforeSubscription`
3. Attempts to `q.dequeueChunk1` which blocks as the queue is empty.
4. `onSubscribe` is called, reads the `Ref` and seeds the state is `Uninitialized`, transitions to `Idle`
5. *crickets*